Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -555,93 +556,102 @@ Ref<WebPageProxy> WebPageProxy::create(PageClient& pageClient, WebProcessProxy&
-    *m_wasRecentlyVisibleActivity = nullptr;
-#endif
+    forEachWebContentProcess([](auto& webProcess, auto) {
+        webProcess.activityState().takeVisibleActivity();
+    });
 }
-void WebPageProxy::ProcessActivityState::takeAudibleActivity()
+void WebPageProxy::takeAudibleActivity()
 {
-    m_isAudibleActivity = m_page.legacyMainFrameProcess().throttler().foregroundActivity("View is playing audio"_s).moveToUniquePtr();
+    forEachWebContentProcess([](auto& webProcess, auto) {
+        webProcess.activityState().takeAudibleActivity();
+    });
 }
-void WebPageProxy::ProcessActivityState::takeCapturingActivity()
+void WebPageProxy::takeCapturingActivity()
 {
-    m_isCapturingActivity = m_page.legacyMainFrameProcess().throttler().foregroundActivity("View is capturing media"_s).moveToUniquePtr();
+    forEachWebContentProcess([](auto& webProcess, auto) {
+        webProcess.activityState().takeCapturingActivity();
+    });
 }
 
-void WebPageProxy::ProcessActivityState::reset()
+void WebPageProxy::resetActivityState()
 {
-    m_isVisibleActivity = nullptr;
-#if PLATFORM(MAC)
-    *m_wasRecentlyVisibleActivity = nullptr;
-#endif
-    m_isAudibleActivity = nullptr;
-    m_isCapturingActivity = nullptr;
-#if PLATFORM(IOS_FAMILY)
-    m_openingAppLinkActivity = nullptr;
-#endif
+    forEachWebContentProcess([](auto& webProcess, auto) {
+        webProcess.activityState().reset();
+    });
 }
 
-void WebPageProxy::ProcessActivityState::dropVisibleActivity()
+void WebPageProxy::dropVisibleActivity()
 {
-#if PLATFORM(MAC)
-    if (WTF::numberOfProcessorCores() > 4)
-        *m_wasRecentlyVisibleActivity = m_page.legacyMainFrameProcess().throttler().backgroundActivity("View was recently visible"_s);
-    else
-        *m_wasRecentlyVisibleActivity = m_page.legacyMainFrameProcess().throttler().foregroundActivity("View was recently visible"_s);
-#endif
-    m_isVisibleActivity = nullptr;
+    forEachWebContentProcess([](auto& webProcess, auto) {
+        webProcess.activityState().dropVisibleActivity();
+    });
 }
 
-void WebPageProxy::ProcessActivityState::dropAudibleActivity()
+void WebPageProxy::dropAudibleActivity()
 {
-    m_isAudibleActivity = nullptr;
+    forEachWebContentProcess([](auto& webProcess, auto) {
+        webProcess.activityState().takeAudibleActivity();
+    });
 }
 
-void WebPageProxy::ProcessActivityState::dropCapturingActivity()
+void WebPageProxy::dropCapturingActivity()
 {
-    m_isCapturingActivity = nullptr;
+    forEachWebContentProcess([](auto& webProcess, auto) {
+        webProcess.activityState().takeCapturingActivity();
+    });
 }
 
-bool WebPageProxy::ProcessActivityState::hasValidVisibleActivity() const
+bool WebPageProxy::hasValidVisibleActivity() const
 {
-    return m_isVisibleActivity && m_isVisibleActivity->isValid();
+    bool hasValidVisibleActivity = true;
+    forEachWebContentProcess([&](auto& webProcess, auto) {
+        hasValidVisibleActivity &= webProcess.activityState().hasValidVisibleActivity();
```

Comment:
We may want these to be or instead of and for these.

Reply:
This needs to be &= because when a new web process is created, this function returns false, so WebPageProxy::updateThrottleState() takes the assertion for the new process. If it was |= the new web process wouldn’t get the assertion.